File: zanata-war/src/main/java/org/zanata/webtrans/shared/validation/action/PrintfVariablesValidation.java
/*
 * Copyright 2011, Red Hat, Inc. and individual contributors
 * as indicated by the @author tags. See the copyright.txt file in the
 * distribution for a full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
package org.zanata.webtrans.shared.validation.action;

import java.util.ArrayList;
import java.util.List;

import org.zanata.webtrans.client.resources.ValidationMessages;
import org.zanata.webtrans.shared.model.ValidationId;
import org.zanata.webtrans.shared.validation.AbstractValidationAction;

import com.google.common.base.Strings;
import com.google.common.collect.Lists;
import com.google.gwt.regexp.shared.MatchResult;
import com.google.gwt.regexp.shared.RegExp;

/**
 * 
 * @author Alex Eng <a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
 * 
 **/
public class PrintfVariablesValidation extends AbstractValidationAction
{
   private static final String GLOBAL_FLAG = "g";

   // derived from translate toolkit printf style variable matching regex. See:
   // http://translate.svn.sourceforge.net/viewvc/translate/src/trunk/translate/filters/checks.py?revision=17978&view=markup
   private static final String VAR_REGEX = "%((?:\\d+\\$|\\(\\w+\\))?[+#-]*(\\d+)?(\\.\\d+)?(hh|h|ll|l|L|z|j|t)?[\\w%])";

   public PrintfVariablesValidation(ValidationId id, ValidationMessages messages)
   {
      this(id, messages.printfVariablesValidatorDesc(), messages);
   }

   public PrintfVariablesValidation(ValidationId id, String description, ValidationMessages messages)
   {
      super(id, description, messages);
   }

   @Override
   public List<String> doValidate(String source, String target)
   {
      ArrayList<String> errors = new ArrayList<String>();

      ArrayList<String> sourceVars = findVars(source);
      ArrayList<String> targetVars = findVars(target);

      String message = findMissingVariables(sourceVars, targetVars);
      if (!Strings.isNullOrEmpty(message))
      {
         errors.add(message);
      }
      message = findAddedVariables(sourceVars, targetVars);
      if (!Strings.isNullOrEmpty(message))
      {
         errors.add(message);
      }

      return errors;
   }

   protected String findMissingVariables(ArrayList<String> sourceVars,
         ArrayList<String> targetVars)
   {
      List<String> missing = listMissing(sourceVars, targetVars);

      if (!missing.isEmpty())
      {
         return getMessages().varsMissing(missing);
      }

      return null;
   }

   protected String findAddedVariables(ArrayList<String> sourceVars,
         ArrayList<String> targetVars)
   {
      // missing from source = added
      List<String> added = listMissing(targetVars, sourceVars);
      if (!added.isEmpty())
      {
         return getMessages().varsAdded(added);
      }

      return null;
   }

   private List<String> listMissing(ArrayList<String> baseVars, ArrayList<String> testVars)
   {
      ArrayList<String> remainingVars = Lists.newArrayList(testVars);

      ArrayList<String> unmatched = Lists.newArrayList();

      for (String var : baseVars)
      {
         if (!remainingVars.remove(var))
         {
            unmatched.add(var);
         }
      }
      return unmatched;
   }

   protected ArrayList<String> findVars(String inString)
   {
      ArrayList<String> vars = new ArrayList<String>();
      // compile each time to reset index
      RegExp varRegExp = RegExp.compile(VAR_REGEX, GLOBAL_FLAG);
      MatchResult result = varRegExp.exec(inString);
      while (result != null)
      {
         vars.add(result.getGroup(0));
         result = varRegExp.exec(inString);
      }
      return vars;
   }
}

File: server/zanata-frontend/src/app/validators/impl/PrintfVariablesValidation.ts
/*
 * Copyright 2018, Red Hat, Inc. and individual contributors
 * as indicated by the @author tags. See the copyright.txt file in the
 * distribution for a full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

 /* tslint:disable:max-line-length*/

import AbstractValidationAction from '../AbstractValidationAction'
import ValidationMessages from '../ValidationMessages'

import MessageFormat from 'intl-messageformat'

/**
 *
 * @author Alex Eng [aeng@redhat.com](mailto:aeng@redhat.com)
 */
class PrintfVariablesValidation extends AbstractValidationAction {
  public id = 'PRINTF_VARIABLES'
  public readonly description: string
  public readonly messages: ValidationMessages
  public readonly label: string

  public sourceExample =
    "value must be between %x and %y"
  public targetExample =
    "value must be between %x and <span class='js-example__target txt--warning'>%z</span>";

  private GLOBAL_FLAG = "g";

  // derived from translate toolkit printf style variable matching regex. See:
  // http://translate.svn.sourceforge.net/viewvc/translate/src/trunk/translate/filters/checks.py?revision=17978&view=markup
  private VAR_REGEX = "%((?:\\d+\\$|\\(\\w+\\))?[+#-]*(\\d+)?(\\.\\d+)?(hh|h|ll|l|L|z|j|t)?[\\w%])"

  constructor(messages: ValidationMessages, locale: string) {
    super(locale)
    this.messages = messages
    this.description = messages.printfVariablesValidatorDesc
    this.label = messages[this.id]
  }

  public doValidate(source: string, target: string): string[] {
    const errors: string[] = []
    const sourceVars = this.findVars(source)
    const targetVars = this.findVars(target)
    let messages = this.findMissingVariables(sourceVars, targetVars)
    if (messages) {
      errors.push(messages)
    }
    messages = this.findAddedVariables(sourceVars, targetVars)
    if (messages) {
      errors.push(messages)
    }
    return errors
  }

  protected findMissingVariables(sourceVars: string[],
    targetVars?: string[]): string {
    const missing = this.listMissing(sourceVars, targetVars)
    return (missing.length > 0)
      ? new MessageFormat(this.messages.varsMissing, this.locale)
        .format({ missing })
      : null
  }

  protected findAddedVariables(sourceVars: string[], targetVars?: string[]): string {
    // missing from source = added
    const added = this.listMissing(targetVars, sourceVars)
    // Push as combined incompleteEntry errors
    return (added.length > 0)
      ?  new MessageFormat(this.messages.varsAdded, this.locale)
        .format({ added })
      : null
  }

  protected findVars(inString: string): string[] {
    const vars: string[] = []
    // compile each time to reset index
    const varRegExp = new RegExp(this.VAR_REGEX, this.GLOBAL_FLAG)
    let result = varRegExp.exec(inString)
    while (result != null) {
      vars.push(result[0])
      result = varRegExp.exec(inString)
    }
    return vars
  }

  private listMissing(baseVars: string[], testVars: string[]): string[] {
      const remainingVars = testVars
      const unmatched: string[] = []
      for (const baseVar of baseVars) {
        const index = remainingVars.indexOf(baseVar)
        if (index !== -1) {
          remainingVars.splice(index, 1);
        } else {
          unmatched.push(baseVar)
        }
      }
      return unmatched
  }
}

export default PrintfVariablesValidation
